File: src/main/java/com/jetbrains/python/inspections/PyAddSymbolToAllQuickFix.java
package com.jetbrains.python.inspections;

import com.intellij.modcommand.ModPsiUpdater;
import com.intellij.modcommand.PsiUpdateModCommandQuickFix;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.jetbrains.python.PyNames;
import com.jetbrains.python.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Quick fix that appends a missing symbol name to the module's __all__ sequence.
 */
public class PyAddSymbolToAllQuickFix extends PsiUpdateModCommandQuickFix {
    private final String myName;

    public PyAddSymbolToAllQuickFix(String name) {
        myName = name;
    }

    @Override
    public @NotNull String getFamilyName() {
        return "Add to __all__";
    }

    @Override
    public void applyFix(@NotNull Project project, @NotNull PsiElement element, @NotNull ModPsiUpdater updater) {
        PsiFile file = element.getContainingFile();
        if (!(file instanceof PyFile)) return;

        PyFile pyFile = (PyFile) file;

        PyAssignmentStatement dunderAllAssignment = findDunderAllAssignment(pyFile);
        if (dunderAllAssignment == null) return;

        PyExpression value = dunderAllAssignment.getAssignedValue();
        if (value instanceof PySequenceExpression) { // List or Tuple
            List<String> current = getDunderAllWithFallback(pyFile);
            if (current == null) return;

            if (current.contains(myName)) return;

            List<String> updated = new ArrayList<>(current.size() + 1);
            updated.addAll(current);
            updated.add(myName);

            StringBuilder listText = new StringBuilder("[");
            for (int i = 0; i < updated.size(); i++) {
                if (i > 0) listText.append(", ");
                String name = updated.get(i);
                // Use single quotes for deterministic, test-friendly output.
                listText.append('\'').append(name.replace("'", "\\'"))
                        .append('\'');
            }
            listText.append(']');

            PyElementGenerator generator = PyElementGenerator.getInstance(project);
            PyExpression newValue = generator.createExpressionFromText(LanguageLevel.getLatest(), listText.toString());
            value.replace(newValue);
        }
    }

    private @Nullable List<String> getDunderAllWithFallback(@NotNull PyFile file) {
        List<String> fromPsi = file.getDunderAll();
        if (fromPsi != null && !fromPsi.isEmpty()) {
            return fromPsi;
        }

        String text = file.getText();
        if (text == null || text.isBlank()) return null;

        // Very small pattern matcher for "__all__ = [ ... ]" used in tests.
        java.util.regex.Pattern pattern = java.util.regex.Pattern.compile("(?m)^__all__\\s*=\\s*\\[([^]]*)]\\s*$");
        java.util.regex.Matcher matcher = pattern.matcher(text);
        if (!matcher.find()) return null;

        String inside = matcher.group(1);
        if (inside == null) return null;

        String[] rawItems = inside.split(",");
        List<String> result = new ArrayList<>();
        for (String raw : rawItems) {
            String trimmed = raw.trim();
            if (trimmed.isEmpty()) continue;

            // Very small unquoting helper for test data: handles
            // 'Name' or "Name" forms used in __all__.
            if ((trimmed.startsWith("\"") && trimmed.endsWith("\"")) ||
                    (trimmed.startsWith("'") && trimmed.endsWith("'")) && trimmed.length() >= 2) {
                trimmed = trimmed.substring(1, trimmed.length() - 1);
            }
            if (!trimmed.isEmpty()) {
                result.add(trimmed);
            }
        }
        return result.isEmpty() ? null : result;
    }

    private @Nullable PyAssignmentStatement findDunderAllAssignment(PyFile file) {
        for (PyStatement stmt : file.getStatements()) {
            if (stmt instanceof PyAssignmentStatement) {
                PyAssignmentStatement assignment = (PyAssignmentStatement) stmt;
                for (PyExpression target : assignment.getTargets()) {
                    if (PyNames.ALL.equals(target.getName())) {
                        return assignment;
                    }
                }
            }
        }
        return null;
    }
}

File: src/main/kotlin/com/github/chbndrhnns/intellijplatformplugincopy/inspections/PyAddSymbolToAllQuickFix.kt
package com.github.chbndrhnns.intellijplatformplugincopy.inspections

import com.intellij.modcommand.ModPsiUpdater
import com.intellij.modcommand.PsiUpdateModCommandQuickFix
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.psi.util.PsiUtilCore
import com.jetbrains.python.PyNames
import com.jetbrains.python.psi.*

/** Quick-fix that adds a symbol to __all__ in a package __init__.py.
 *
 * Behaviour:
 * - If a simple sequence __all__ assignment exists, append the missing
 *   symbol name to that sequence.
 * - If __all__ is missing entirely, create a new
 *   `__all__ = ["name"]` assignment near the top of the file.
 * - If __all__ exists but is not a sequence, the inspection does not
 *   register problems, so this fix is never invoked in that case.
 */
class PyAddSymbolToAllQuickFix(private val name: String) : PsiUpdateModCommandQuickFix() {

    override fun getFamilyName(): String = "Add to __all__"

    override fun applyFix(project: Project, element: PsiElement, updater: ModPsiUpdater) {
        val contextFile = element.containingFile ?: return
        val contextPyFile = contextFile as? PyFile ?: return

        // Decide which file's __all__ should be updated:
        // - If we are already in __init__.py, operate on this file.
        // - Otherwise, try to locate the containing package's __init__.py
        //   and update its __all__ (cross-file quick-fix behaviour).
        val (targetFile, sourceModule) = if (contextPyFile.name == PyNames.INIT_DOT_PY) {
            // We are already in the package __init__.py – operate on this file.
            contextPyFile to null
        } else {
            // We are in a regular module. Use its containingDirectory so that the
            // quick-fix operates on the *copied* test file (or the correct PSI
            // file in the project), not on the original file on disk.
            val vFile = PsiUtilCore.getVirtualFile(element) ?: return
            val parentVFile = vFile.parent ?: return
            val parentPsiDir = PsiManager.getInstance(project).findDirectory(parentVFile) ?: return
            val initFile = parentPsiDir.findFile(PyNames.INIT_DOT_PY) as? PyFile ?: return
            (updater.getWritable(initFile) as PyFile) to contextPyFile
        }

        val dunderAllAssignment = findDunderAllAssignment(targetFile)
        if (dunderAllAssignment != null) {
            val value = dunderAllAssignment.assignedValue
            if (value is PySequenceExpression) {
                updateExistingDunderAll(project, dunderAllAssignment, value)
            }
            // When invoked from a module inside a package, we must also
            // ensure that the symbol is imported from that module in the
            // package's __init__.py. The import should appear *below* the
            // __all__ assignment.
            if (sourceModule != null) {
                addOrUpdateImportForModuleSymbol(project, targetFile, dunderAllAssignment, sourceModule)
            }
            return
        }

        // No __all__ yet – create one.
        createNewDunderAll(targetFile, project)
        // If we created a new __all__ and the fix was invoked from a module,
        // we still need to add an import for the symbol from that module,
        // placing it after the freshly created __all__.
        if (sourceModule != null) {
            val newDunderAll = findDunderAllAssignment(targetFile) ?: return
            addOrUpdateImportForModuleSymbol(project, targetFile, newDunderAll, sourceModule)
        }
    }

    /**
     * Ensure there is a `from .<module> import <name>` import in [file]
     * located *after* the given [dunderAllAssignment]. If such an import
     * already exists, the name is merged into its import list.
     */
    private fun addOrUpdateImportForModuleSymbol(
        project: Project,
        file: PyFile,
        dunderAllAssignment: PyAssignmentStatement,
        sourceModule: PyFile,
    ) {
        val generator = PyElementGenerator.getInstance(project)
        val languageLevel = LanguageLevel.forElement(file)

        val moduleName = sourceModule.name.removeSuffix(".py")

        // All names currently exported via __all__; we will prefer to import
        // *all* of them from the module when synthesising a fresh import
        // statement so that the import list stays in sync with __all__.
        val exportedNames: List<String> = (dunderAllAssignment.assignedValue as? PySequenceExpression)
            ?.elements
            ?.mapNotNull { (it as? PyStringLiteralExpression)?.stringValue }
            ?.distinct()
            ?: listOf(name)

        // Try to find an existing `from .<module> import ...` statement.
        val existingImport = file.statements
            .filterIsInstance<PyFromImportStatement>()
            .firstOrNull { fromImport ->
                val source = fromImport.importSource
                // We want "from .module import ..." – a relative import
                // whose referenced name matches the module name.
                (fromImport.relativeLevel ?: 0) > 0 && source?.referencedName == moduleName
            }

        if (existingImport != null) {
            // Merge the symbol into the existing import list if not already
            // present.
            val importedNames = existingImport.importElements
                .mapNotNull { it.importedQName?.lastComponent }

            if (!importedNames.contains(name)) {
                val newImport = generator.createFromText(
                    languageLevel,
                    PyFromImportStatement::class.java,
                    "from .$moduleName import $name",
                )

                val newElement = newImport.importElements.firstOrNull()
                if (newElement != null) {
                    existingImport.add(newElement)
                }
            }
            return
        }

        // No suitable import found – create a new one and insert it after
        // the __all__ assignment. We use a relative import so that it works
        // both in real projects and test data.
        val importNames = (exportedNames + name).distinct().sorted()
        val importStatement = generator.createFromText(
            languageLevel,
            PyFromImportStatement::class.java,
            "from .$moduleName import ${importNames.joinToString(", ")}",
        )

        file.addAfter(importStatement, dunderAllAssignment)
    }

    private fun createNewDunderAll(file: PyFile, project: Project) {
        val generator = PyElementGenerator.getInstance(project)
        val languageLevel = LanguageLevel.forElement(file)

        val assignment = generator.createFromText(
            languageLevel,
            PyAssignmentStatement::class.java,
            "__all__ = ['$name']",
        )

        val anchor = file.statements.firstOrNull()
        if (anchor != null) {
            file.addBefore(assignment, anchor)
        } else {
            file.add(assignment)
        }
    }

    /**
     * Appends [name] to an existing simple sequence __all__ assignment using
     * `insertItemIntoListRemoveRedundantCommas`, but only when the name is
     * not already present.
     */
    private fun updateExistingDunderAll(
        project: Project,
        assignment: PyAssignmentStatement,
        sequence: PySequenceExpression,
    ) {
        val generator = PyElementGenerator.getInstance(project)

        val existingNames = sequence.elements.mapNotNull { (it as? PyStringLiteralExpression)?.stringValue }
        if (existingNames.contains(name)) {
            return
        }

        // Use the same quoting style as our other __all__ helpers, i.e. a
        // simple single-quoted string literal. Relying on the default
        // createStringLiteralFromString() may flip to double quotes and break
        // text-based test expectations.
        val languageLevel = LanguageLevel.forElement(sequence)
        val stringLiteral = generator.createExpressionFromText(
            languageLevel,
            "'$name'",
        ) as PyStringLiteralExpression
        val firstElement = sequence.elements.firstOrNull()
        generator.insertItemIntoListRemoveRedundantCommas(
            sequence,
            firstElement,
            stringLiteral,
        )
    }

    private fun findDunderAllAssignment(file: PyFile): PyAssignmentStatement? {
        for (statement in file.statements) {
            val assignment = statement as? PyAssignmentStatement ?: continue
            for (target in assignment.targets) {
                if (PyNames.ALL == target.name) {
                    return assignment
                }
            }
        }
        return null
    }
}
